### ISSUE TITLE
Explore Android native Spell-Checker

### ISSUE THREAD

**Issue by menny** [AnySoftKeyboard/AnySoftKeyboard#1215]
https://android.googlesource.com/platform/development/+/master/samples/SpellChecker/HelloSpellChecker/src/com/example/android/hellospellchecker/HelloSpellCheckerActivity.java

**Comment by mondragondavid** [comment]
Yes, users have been very unsatisfied with the current Spell Checker.  Also, we've looked at the code to try to improve this but are struggling to get it to work.

**Comment by jwsp1** [comment]
This would be a major improvement since lots of words are missing in language dictonaries. For example, I had to add hundreds of well-known words to complete German dictonary...

**Comment by ildar** [comment]
.. or alternative approach possible: using hunspell (standard in Linux and
Libre/OpenOffice).

ref: https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/1229

**Comment by dvaerum** [comment]
I am not sure if I am reading this correct, but does the spell checker in AnySoftKeyboard not use the native spell checker in android?

**Comment by mirfatif** [comment]
> I am not sure if I am reading this correct, but does the spell checker in AnySoftKeyboard not use the native spell checker in android?

ASK should not rely on AOSP's spell checker but itself provide a spell checker like Gboard does i.e. red underline the wrongly spelled words.

**Comment by dvaerum** [comment]
I would love to be able to use the LanguageTool as the spell checker for AnySoftKeyboard and from what I understand that will be possible if using the Android native Spell-Checker

**Comment by mirfatif** [comment]
> I would love to be able to use the LanguageTool as the spell checker for AnySoftKeyboard and from what I understand that will be possible if using the Android native Spell-Checker

How would Android's native spell checker use LanguageTool? You need to install some third party app which is built on LanguageTool APIs and Android's spell checker framework. Here is one such app: https://github.com/Softcatala/LanguageToolAndroidService

**Comment by dvaerum** [comment]
> How would Android's native spell checker use LanguageTool? Native spell checker has its own limited set of words. You need to install some third party app which (can be ASK if it) is built on LanguageTool APIs and Android's spell checker framework. Here is one such app: https://github.com/Softcatala/LanguageToolAndroidService

Yes, you need a third-party app, but if I understand it correctly if AnySoftKeyboard support the Android native Spell-Checker you can set LanguageTool (or some other 3rd-party software) as your default spell-checker in Android and that way it would work in AnySoftKeyboard. Again, if I understand in correctly :sweat_smile:

**Comment by mirfatif** [comment]
> Yes, you need a third-party app, but if I understand it correctly if AnySoftKeyboard support the Android native Spell-Checker you can set LanguageTool (or some other 3rd-party software) as your default spell-checker in Android and that way it would work in AnySoftKeyboard. Again, if I understand in correctly sweat_smile

If you want to use LanguageTool or any other third party spell checker which is available as an Android app, you are all set to go. There's nothing needed from ASK side. Spell checker works independent of the keyboard being used (not sure though if keyboard has some involvement). EditText view (or whatever component where you type text) of an app interact with AOSP native spell checker (or whatever spell checker you installed as a third party app and selected in device settings) to highlight misspelled words. Right now you install your favorite spell checker and it'd be working.

What's being discussed here is that ASK should itself provide spell checking service (in addition to being a keyboard) so that the pre-built and user generated dictionary of every language may also be used for spell checking (in addition to auto-correction and next word suggestion). In this ways user generated dictionary can be backed up and restored easily, without waiting for a new dictionary to be built based on your typing experience. That's what Gboard does.

**Comment by dvaerum** [comment]
> What's being discussed here is that ASK should itself provide spell checking service (in addition to being a keyboard) so that the pre-built and user generated dictionary of every language may also be used for spell checking (in addition to auto-correction and next word suggestion). In this ways user generated dictionary can be backed up and restored easily, without waiting for a new dictionary to be built based on your typing experience. That's what Gboard does

Thanks for clarifying, my misunderstanding :sweat_smile: 

> If you want to use LanguageTool or any other third party spell checker which is available as an Android app, you are all set to go. There's nothing needed from ASK side. Spell checker works independent of the keyboard being used (not sure though if keyboard has some involvement). EditText view (or whatever component where you type text) of an app interact with AOSP native spell checker (or whatever spell checker you installed as a third party app and selected in device settings) to highlight misspelled words. Right now you install your favorite spell checker and it'd be working.

Guess I need to try this out, again :smiley:
